sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObject.java

```diff
@@ -13,236 +13,11 @@
-
-        handlersCache.put(sqlObjectType, handlers);
-
-        return handlers;
-    }
-
-    @SafeVarargs
-    private static void attachClose(Map<Method, Handler> handlers, Class<? extends AutoCloseable>... classes)
-    {
-        for (Class<?> klass : classes) {
-            Method closeMethod;
-            try {
-                closeMethod = klass.getMethod("close");
-            } catch (NoSuchMethodException | SecurityException e) {
-                throw new UnableToCreateSqlObjectException("Could not find AutoCloseable#close");
-            }
-            handlers.put(closeMethod, new CloseHandler());
-        }
-    }
-
-
-    private final Map<Method, Handler> handlers;
-    private final HandleDing           ding;
-
-    SqlObject(Map<Method, Handler> handlers, HandleDing ding)
-    {
-        this.handlers = handlers;
-        this.ding = ding;
-    }
-
-    public Object invoke(Object proxy, Method method, Object[] args, MethodProxy mp) throws Throwable
-    {
-        final Handler handler = handlers.get(method);
-
-        // If there is no handler, pretend we are just an Object and don't open a connection (Issue #82)
-        if (handler == null) {
-            return mp.invokeSuper(proxy, args);
-        }
-
-        Throwable doNotMask = null;
-        try {
-            ding.retain(method.toString());
-            return handler.invoke(ding, proxy, args, mp);
-        }
-        catch (Throwable e) {
-            doNotMask = e;
-            throw e;
-        }
-        finally {
-            try {
-                ding.release(method.toString());
-            }
-            catch (Throwable e) {
-                if (doNotMask==null) {
-                    throw e;
-                }
-            }
-        }
-    }
-
-    public static void close(Object sqlObject)
-    {
-        if (!(sqlObject instanceof CloseMe)) {
-            throw new IllegalArgumentException(sqlObject + " is not a sql object");
-        }
-        ((CloseMe) sqlObject).close();
-    }
-
-    static String getSql(SqlCall q, Method m)
-    {
-        if (SqlQuery.DEFAULT_VALUE.equals(q.value())) {
-            return m.getName();
-        }
-        else {
-            return q.value();
-        }
-    }
-
-    static String getSql(SqlQuery q, Method m)
-    {
-        if (SqlQuery.DEFAULT_VALUE.equals(q.value())) {
-            return m.getName();
-        }
-        else {
-            return q.value();
-        }
-    }
-
-    static String getSql(SqlUpdate q, Method m)
-    {
-        if (SqlQuery.DEFAULT_VALUE.equals(q.value())) {
-            return m.getName();
-        }
-        else {
-            return q.value();
-        }
-    }
-
-    static String getSql(SqlBatch q, Method m)
-    {
-        if (SqlQuery.DEFAULT_VALUE.equals(q.value())) {
-            return m.getName();
-        }
-        else {
-            return q.value();
-        }
+public class SqlObject implements ExtensionConfig<SqlObject> {
```

Comment:
I thought about that, but decided against it for forward compatibility. Over time a user would expect the config class to stay in the same place, but accumulate more options.

I'm afraid if we provide a `NoConfig` we'd be encouraging extension authors into a situation where they have to break backward compatibility in order to make their extensions configurable.

I'm still debating whether having the configuration class separate from the extension factory is really necessary.